code/game/machinery/dance_machine.dm

```diff
@@ -0,0 +1,505 @@
+		if (!M)
+			return
+		if (i<31)
+			initial_matrix = matrix(M.transform)
+			initial_matrix.Translate(0,1)
+			animate(M, transform = initial_matrix, time = 1, loop = 0)
+		if (i>30)
+			initial_matrix = matrix(M.transform)
+			initial_matrix.Translate(0,-1)
+			animate(M, transform = initial_matrix, time = 1, loop = 0)
+		M.setDir(turn(M.dir, 90))
+		switch (M.dir)
+			if (NORTH)
+				initial_matrix = matrix(M.transform)
+				initial_matrix.Translate(0,3)
+				animate(M, transform = initial_matrix, time = 1, loop = 0)
+			if (SOUTH)
+				initial_matrix = matrix(M.transform)
+				initial_matrix.Translate(0,-3)
+				animate(M, transform = initial_matrix, time = 1, loop = 0)
+			if (EAST)
+				initial_matrix = matrix(M.transform)
+				initial_matrix.Translate(3,0)
+				animate(M, transform = initial_matrix, time = 1, loop = 0)
+			if (WEST)
+				initial_matrix = matrix(M.transform)
+				initial_matrix.Translate(-3,0)
+				animate(M, transform = initial_matrix, time = 1, loop = 0)
+		sleep (1)
+	animate(M, transform = null, time = 1, loop = 0)
+
+/mob/living/carbon/proc/dancey()
+	var/matrix/initial_matrix = matrix(transform)
+	spawn (0)
+		if (src)
+			animate(src, transform = matrix(180, MATRIX_ROTATE), time = 1, loop = 0)
+			initial_matrix = matrix(transform)
+		sleep (70)
+		if (src)
+			animate(src, transform = null, time = 1, loop = 0)
+	for (var/i = 0, i < 60, i++)
+		if (!src)
+			return
+		if (i<31)
+			initial_matrix = matrix(transform)
+			initial_matrix.Translate(0,1)
+			transform = initial_matrix
+			animate(src, transform, time = 1, loop = 0)
+		if (i>30)
+			initial_matrix = matrix(transform)
+			initial_matrix.Translate(0,-1)
+			transform = initial_matrix
+			animate(src, transform, time = 1, loop = 0)
+		setDir(turn(src.dir, 90))
+		switch (dir)
+			if (NORTH)
+				initial_matrix = matrix(transform)
+				initial_matrix.Translate(0,3)
+				animate(src, transform, time = 1, loop = 0)
+			if (SOUTH)
+				initial_matrix = matrix(transform)
+				initial_matrix.Translate(0,-3)
+				animate(src, transform, time = 1, loop = 0)
+			if (EAST)
+				initial_matrix = matrix(transform)
+				initial_matrix.Translate(3,0)
+				animate(src, transform, time = 1, loop = 0)
+			if (WEST)
+				initial_matrix = matrix(transform)
+				initial_matrix.Translate(-3,0)
+				animate(src, transform, time = 1, loop = 0)
+		sleep (1)
+	animate(src, transform = null, time = 1, loop = 0)
+
+
+/obj/machinery/disco/proc/dance_over()
+	for(var/obj/item/device/flashlight/spotlight/SL in spotlights)
+		qdel(SL)
+	spotlights.Cut()
+	for(var/obj/effect/overlay/sparkles/SP in sparkles)
+		qdel(SP)
+	sparkles.Cut()
+	rangers.Cut()
+	for(var/mob/living/L in listeners)
+		if(!L || !L.client)
+			continue
+		L.stop_sound_channel(CHANNEL_JUKEBOX)
+	listeners.Cut()
+
+
+/obj/machinery/disco/process()
+	if(charge<35)
+		charge += 1
+	if(world.time < stop && active)
+		rangers = list()
+		for(var/mob/living/M in range(9,src))
+			rangers += M
+			if(!(M in listeners))
```

Comment:
that is correct, yes.
```list[SOME_KEY] = SOME_VALUE```` inserts ```SOME_KEY``` into the list, so if you use it like a normal list, you get keys, but you can use keys to VERY quickly access the ```SOME_VALUE```s, in this case the ```TRUE``` were using is just because we want the ```if()``` to pass if there's a value at the key.